### PR TITLE
[jit] Add `save()` to `torch._C.Function`

### LINKED ISSUE #20017: torch.jit.trace returns unwrapped C type
### With pytorch 1.1.0

```
import torch
def f(x):
  return x*2
z = torch.jit.trace(f, (torch.zeros(10),))
print(type(z))
```

Returns `torch._C.Function` and cannot be saved:

`Traceback (most recent call last):
  File "1.py", line 7, in <module>
    torch.jit.save(z, "filename")
  File "/home/vitalyf/local/miniconda/envs/wp_1/lib/python3.7/site-packages/torch/jit/__init__.py", line 198, in save
    m.save(f, _extra_files=_extra_files)
AttributeError: 'torch._C.Function' object has no attribute 'save'
`

### Expected as in pytorch 1.0.1

```
import torch
def f(x):
  return x*2
z = torch.jit.trace(f, (torch.zeros(10),))
print(type(z))
```

Returns `<class 'torch.jit.TopLevelTracedModule'>`

### PR DESCRIPTION
**Pull request by driazati**
Fixes #20017

This wraps the `torch._C.Function` currently returned from `torch.jit.script` and `torch.jit.trace` in a `ScriptFunction` and `TracedFunction` respectively, both of which are just wrappers to hold the function.

Differential Revision: [D15403161](https://our.internmc.facebook.com/intern/diff/15403161/)

### CODE REVIEW

File: torch/csrc/jit/script/init.cpp [hunk 1]
```diff
@@ -476,6 +476,35 @@ void initJitScriptBindings(PyObject* module) {
...
           })
+      .def(
+          "save",
+          [](std::shared_ptr<Function> self,
+             const std::string& filename,
+             const ExtraFilesMap& _extra_files = ExtraFilesMap()) {
+            Module module;
+            // Make a graph with a fake self argument
+            auto graph = self->graph()->copy();
+            graph->insertInput(0, "self");
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: The type of input 0 is not set here (it is also never checked, which is why it doesn't fail).

File: torch/csrc/jit/script/init.cpp [hunk 2]
```diff
@@ -476,6 +476,35 @@ void initJitScriptBindings(PyObject* module) {
...
+          },
+          py::arg("filename"),
+          py::arg("_extra_files") = ExtraFilesMap())
+      .def(
+          "save_to_buffer",
+          [](std::shared_ptr<Function> self,
+             const ExtraFilesMap& _extra_files = ExtraFilesMap()) {
+            std::ostringstream buf;
+            Module module;
+            auto graph = self->graph()->copy();
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: Can we put the implementation in Function::save, and avoid duplicating the logic between save and save_to_buffer?

**Review by zdevito — APPROVED** [review]
Looks good, minor comments.

**Comment by apaszke** [comment]
Does `load(save(f))` give you back a function (as it should) or a module (as it seems to do from what I see)?

**Comment by driazati** [comment]
It does return a module for now, our serialization doesn't know about functions afaik and we wanted to at least get the save/load functionality back before fixing that

**Pull request closed without merging**